src/game/client/water/w3dwater.cpp

```diff
@@ -0,0 +1,2327 @@
+        m_pDev->SetTextureStageState(0, D3DTSS_TEXTURETRANSFORMFLAGS, D3DTTFF_DISABLE);
+        m_pDev->SetTextureStageState(0, D3DTSS_TEXCOORDINDEX, 0);
+        m_pDev->SetTextureStageState(1, D3DTSS_TEXTURETRANSFORMFLAGS, D3DTTFF_DISABLE);
+        m_pDev->SetTextureStageState(1, D3DTSS_TEXCOORDINDEX, 1);
+        m_pDev->SetRenderState(D3DRS_ZWRITEENABLE, TRUE);
+        m_pDev->SetTextureStageState(1, D3DTSS_ADDRESSU, D3DTADDRESS_WRAP);
+        m_pDev->SetTextureStageState(1, D3DTSS_ADDRESSV, D3DTADDRESS_WRAP);
+        m_pDev->SetRenderState(D3DRS_WRAP0, 0);
+        m_pDev->SetTextureStageState(0, D3DTSS_COLOROP, D3DTOP_DISABLE);
+        m_pDev->SetTextureStageState(0, D3DTSS_ALPHAOP, D3DTOP_DISABLE);
+        m_pDev->SetTextureStageState(1, D3DTSS_COLOROP, D3DTOP_DISABLE);
+        m_pDev->SetTextureStageState(1, D3DTSS_ALPHAOP, D3DTOP_DISABLE);
+        m_pDev->SetTextureStageState(2, D3DTSS_COLOROP, D3DTOP_DISABLE);
+        m_pDev->SetTextureStageState(2, D3DTSS_ALPHAOP, D3DTOP_DISABLE);
+        DX8Wrapper::Set_DX8_Transform(D3DTS_VIEW, (Matrix4 &)m2);
+        DX8Wrapper::Set_DX8_Transform(D3DTS_PROJECTION, (Matrix4 &)m);
+        m_pDev->SetPixelShader(0);
+        m_pDev->SetVertexShader(DX8_FVF_XYZDUV1);
+        DX8Wrapper::Invalidate_Cached_Render_States();
+
+        if (g_theTerrainRenderObject->Get_Shroud()) {
+            W3DShaderManager::Set_Texture(0, g_theTerrainRenderObject->Get_Shroud()->Get_Shroud_Texture());
+            W3DShaderManager::Set_Shader(W3DShaderManager::ST_SHROUD_TEXTURE, 0);
+            m_pDev->SetStreamSource(0, m_vertexBufferD3D, 24);
+            m_pDev->SetIndices(m_indexBufferD3D, 0);
+
+            for (int i = ((box.m_center.Y - box.m_extent.Y) / 560.0f);
+                 box.m_center.Y + box.m_extent.Y > (float)(14 * i) * 40.0f;
+                 i++) {
+                for (int j = ((box.m_center.X - box.m_extent.X) / 560.0f);
+                     box.m_center.X + box.m_extent.X > (float)(14 * j) * 40.0f;
+                     j++) {
+                    D3DXMATRIX m10;
+                    m10.m[3][0] = (float)(14 * j) * 40.0f;
+                    m10.m[3][2] = (float)(14 * i) * 40.0f;
+                    D3DXMatrixMultiply(&m10, &m6, &m3);
+                    DX8Wrapper::Set_DX8_Transform(D3DTS_WORLD, (Matrix4 &)m10);
+                    m_pDev->DrawIndexedPrimitive(D3DPT_TRIANGLESTRIP, 0, m_numVertices, 0, m_numIndices);
+                }
+            }
+
+            W3DShaderManager::Reset_Shader(W3DShaderManager::ST_SHROUD_TEXTURE);
+        }
+    }
+#endif
+}
+
+void WaterRenderObjClass::Render_Water()
+{
+    for (PolygonTrigger *i = PolygonTrigger::Get_First_Polygon_Trigger(); i; i = i->Get_Next()) {
+        if (i->Is_Water_Area()) {
+            if (i->Get_Num_Points() > 2) {
+                if (i->Is_River()) {
+                    Draw_River_Water(i);
+                } else {
+                    for (int j = 1; j < i->Get_Num_Points() - 1; j++) {

```

Comment:
original increments by 2?, this whole function feels off...

Reply:
Fixed.